homeassistant/components/media_player/frontier_silicon.py

```diff
@@ -0,0 +1,245 @@
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/media_player.frontier_silicon/
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.media_player import (
+    SUPPORT_NEXT_TRACK, SUPPORT_PAUSE, SUPPORT_PREVIOUS_TRACK, SUPPORT_SEEK,
+    SUPPORT_PLAY_MEDIA, SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET,
+    SUPPORT_VOLUME_STEP, SUPPORT_STOP, SUPPORT_TURN_OFF, SUPPORT_TURN_ON,
+    SUPPORT_PLAY, SUPPORT_SELECT_SOURCE, MediaPlayerDevice, PLATFORM_SCHEMA)
+from homeassistant.const import (
+    STATE_OFF, STATE_PLAYING, STATE_PAUSED, STATE_UNKNOWN,
+    CONF_HOST, CONF_PORT, CONF_PASSWORD)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['fsapi==0.0.7']
+
+_LOGGER = logging.getLogger(__name__)
+
+SUPPORT_FRONTIER_SILICON = SUPPORT_PAUSE | SUPPORT_VOLUME_SET | \
+    SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_STEP | \
+    SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK | SUPPORT_SEEK | \
+    SUPPORT_PLAY_MEDIA | SUPPORT_PLAY | SUPPORT_STOP | SUPPORT_TURN_ON | \
+    SUPPORT_TURN_OFF | SUPPORT_SELECT_SOURCE
+
+DEFAULT_PORT = 80
+DEFAULT_PASSWORD = '1234'
+DEVICE_URL = 'http://{0}:{1}/device'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_PASSWORD, default=DEFAULT_PASSWORD): cv.string,
+})
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the Frontier Silicon platform."""
+    import requests
+
+    if discovery_info is not None:
+        add_devices(
+            [FSAPIDevice(discovery_info, DEFAULT_PASSWORD)],
+            update_before_add=True)
+        return True
+
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    password = config.get(CONF_PASSWORD)
+
+    try:
+        add_devices(
+            [FSAPIDevice(DEVICE_URL.format(host, port), password)],
+            update_before_add=True)
+        _LOGGER.debug('FSAPI device %s:%s -> %s', host, port, password)
+        return True
+    except requests.exceptions.RequestException:
+        _LOGGER.error('Could not add the FSAPI device at %s:%s -> %s',
+                      host, port, password)
+
+    return False
+
+
+class FSAPIDevice(MediaPlayerDevice):
+    """Representation of a Frontier Silicon device on the network."""
+
+    def __init__(self, device_url, password):
+        """Initialize the Frontier Silicon API device."""
+        self._device_url = device_url
+        self._password = password
+        self._state = STATE_UNKNOWN
+
+        self._name = None
+        self._title = None
+        self._mute = None
+        self._source = None
+        self._source_list = None
+        self._media_image_url = None
+
+    def get_fs(self):
+        """Create a fsapi session."""
+        from fsapi import FSAPI
+
+        return FSAPI(self._device_url, self._password)
```

Comment:
Hi, thanks for the input. I do appreciate it. 
There is a single session controlling a frontier silicon device. Once a new connection has been established - the previous session will be invalidated. This makes sense as an Internet Radio can play a single station, song, etc at a time. The reason I always recreate the reference when 'update' is being called, is to ensure I have a valid session. Otherwise I would always have to try-catch-recreate in case someone has used the the Undock app (i.e firmware update, etc.) in between and has destroyed the __init__ session. I am not a Python developer but usually a http request (once every 30 second in the local network) does not cause much overhead. In my experience this pattern is much more robust and  maintainable than posting a request, catching an exception, then eventually recreating the instance and repeating the request. Still I am not a Python developer and I am sure this not the first platform in homeassistant handling this issue. If you have any suggestions, please comment here. I've used the KISS solution here. I would implement an async component as soon as I find time, the fsapi is the bottleneck in this case but I 'll fix it as soon as I can.